components/zcash_note_encryption/src/lib.rs

```diff
@@ -1,8 +1,464 @@
+    fn extract_memo(&self, plaintext: &[u8]) -> Self::Memo;
+
+    fn extract_pk_d(
+        out_plaintext: &[u8; OUT_CIPHERTEXT_SIZE],
+    ) -> Option<Self::DiversifiedTransmissionKey>;
+
+    fn extract_esk(out_plaintext: &[u8; OUT_CIPHERTEXT_SIZE]) -> Option<Self::EphemeralSecretKey>;
+}
+
+pub trait ShieldedOutput<'a, D: Domain> {
+    fn ivk(&'a self) -> &'a D::IncomingViewingKey;
+    fn epk(&'a self) -> &'a D::EphemeralPublicKey;
+    fn cmstar(&'a self) -> &'a D::ExtractedCommitment;
+}
+
+/// A struct containing context required for encrypting Sapling and Orchard notes.
+///
+/// This struct provides a safe API for encrypting Sapling and Orchard notes. In particular, it
+/// enforces that fresh ephemeral keys are used for every note, and that the ciphertexts are
+/// consistent with each other.
+///
+/// Implements section 4.17.1 of the Zcash Protocol Specification.
+/// NB: the example code is only covering the pre-Canopy case.
+///
+/// # Examples
+///
+/// ```
+/// extern crate ff;
+/// extern crate rand_core;
+/// extern crate zcash_primitives;
+///
+/// use ff::Field;
+/// use rand_core::OsRng;
+/// use zcash_primitives::{
+///     consensus::TestNetwork,
+///     memo::MemoBytes,
+///     sapling::{
+///         keys::{OutgoingViewingKey, prf_expand},
+///         note_encryption::{sapling_note_encryption},
+///         Diversifier, PaymentAddress, Rseed, ValueCommitment
+///     },
+/// };
+///
+/// let mut rng = OsRng;
+///
+/// let diversifier = Diversifier([0; 11]);
+/// let pk_d = diversifier.g_d().unwrap();
+/// let to = PaymentAddress::from_parts(diversifier, pk_d).unwrap();
+/// let ovk = Some(OutgoingViewingKey([0; 32]));
+///
+/// let value = 1000;
+/// let rcv = jubjub::Fr::random(&mut rng);
+/// let cv = ValueCommitment {
+///     value,
+///     randomness: rcv.clone(),
+/// };
+/// let rcm = jubjub::Fr::random(&mut rng);
+/// let note = to.create_note(value, Rseed::BeforeZip212(rcm)).unwrap();
+/// let cmu = note.cmu();
+///
+/// let mut enc = sapling_note_encryption::<_, TestNetwork>(ovk, note, to, MemoBytes::empty(), &mut rng);
+/// let encCiphertext = enc.encrypt_note_plaintext();
+/// let outCiphertext = enc.encrypt_outgoing_plaintext(&cv.commitment().into(), &cmu, &mut rng);
+/// ```
+pub struct NoteEncryption<D: Domain> {
+    epk: D::EphemeralPublicKey,
+    esk: D::EphemeralSecretKey,
+    note: D::Note,
+    to: D::Recipient,
+    memo: D::Memo,
+    /// `None` represents the `ovk = ⊥` case.
+    ovk: Option<D::OutgoingViewingKey>,
+}
+
+impl<D: Domain> NoteEncryption<D> {
+    /// Construct a new note encryption context for the specified note,
+    /// recipient, and memo.
+    pub fn new(
+        ovk: Option<D::OutgoingViewingKey>,
+        note: D::Note,
+        to: D::Recipient,
+        memo: D::Memo,
+    ) -> Self {
+        let esk = D::derive_esk(&note).expect("ZIP 212 is active.");
+        Self::new_with_esk(esk, ovk, note, to, memo)
+    }
+
+    /// For use only with Sapling.
```

Comment:
We don't need to be able to generate pre-ZIP-212 ciphertexts, so I think there's no reason for this to be exposed. It can just be part of `new`.

Reply:
The reason to generate pre-zip-212 ciphertexts is to allow tests to continue to cover pre-zip-212 transaction decryption.

Reply:
Ok, that makes sense.